sky/skylet/providers/kubernetes/node_provider.py

```diff
@@ -225,6 +233,129 @@ def _raise_pod_scheduling_errors(self, new_nodes):
+                        all_pods_scheduled = False
+                        break
+
+                    for container_status in pod.status.container_statuses:
+                        # If the container wasn't in 'ContainerCreating'
+                        # state, then we know pod wasn't scheduled or
+                        # had some other error, such as image pull error.
+                        # See list of possible reasons for waiting here:
+                        # https://stackoverflow.com/a/57886025
+                        waiting = container_status.state.waiting
+                        if waiting and waiting.reason != 'ContainerCreating':
+                            all_pods_scheduled = False
+                            break
+
+            if all_pods_scheduled:
+                return
+            time.sleep(1)
+
+        # Handle pod scheduling errors
+        try:
+            self._raise_pod_scheduling_errors(new_nodes)
+        except config.KubernetesError:
+            raise
+        except Exception as e:
+            raise config.KubernetesError(
+                'An error occurred while trying to fetch the reason '
+                'for pod scheduling failure. '
+                f'Error: {common_utils.format_exception(e)}') from None
+
+    def _wait_for_pods_to_run(self, new_nodes):
+        """Wait for pods and their containers to be ready.
+        
+        Pods may be pulling images or may be in the process of container
+        creation.
+        """
+        while True:
+            all_pods_running = True
+            # Iterate over each pod to check their status
+            for node in new_nodes:
+                pod = kubernetes.core_api().read_namespaced_pod(
+                    node.metadata.name, self.namespace)
+
+                # Continue if pod and all the containers within the
+                # pod are succesfully created and running.
+                if pod.status.phase == 'Running' \
+                    and all([container.state.running
+                             for container in pod.status.container_statuses]):
+                    continue
+
+                all_pods_running = False
+                if pod.status.phase == 'Pending':
+                    # Iterate over each container in pod to check their status
+                    for container_status in pod.status.container_statuses:
+                        waiting = container_status.state.waiting
+                        if waiting and (waiting.reason == 'ErrImagePull' or
+                                        waiting.reason == 'ImagePullBackOff'):
+                            raise config.KubernetesError(
+                                'Failed to pull docker image while '
+                                'launching the node. Please check '
+                                'your network connection. Error details: '
+                                f'{container_status.state.waiting.message}.')
```

Comment:
Shouldn't this check and raise be moved to after L265, where we already have similar checks in place? It seems to make more sense to have all waiting related errors handled at one place, and this method should be relegated to simply be a wait loop.

Reply:
We currently have two methods where `_wait_for_pods_to_schedule` waits until all the pods are scheduled and `_wait_for_pods_to_run` waits for all the pods and containers to be in `Running` state. We separated these two methods as the pods being scheduled shouldn't take longer than `self.timeout`, 10s, but waiting for the pods and containers to be in `Running` state after scheduled can take much longer due to the varying image size.

The checks you mentioned in this comment in `_wait_for_pods_to_run` checks the error that get raised after the pods are scheduled. The waiting checks being done from `_wait_for_pods_to_schedule` and `_wait_for_pods_to_run` are different kinds of checks, so they should be separated. Please correct me if I'm missing anything!

Reply:
Update: I realize now that the following check from `_wait_for_pods_to_schedule` should actually be done after the pods are scheduled and doesn't need to be checked from `_wait_for_pods_to_schedule`. `waiting.reason` can be set to `'ContainerCreating'` only after the pods have already been scheduled, so checking if the pods reached `ContainerCreating` state should be placed in `_wait_for_pods_to_run`.
```
                    for container_status in pod.status.container_statuses:
                        # If the container wasn't in 'ContainerCreating'
                        # state, then we know pod wasn't scheduled or
                        # had some other error, such as image pull error.
                        # See list of possible reasons for waiting here:
                        # https://stackoverflow.com/a/57886025
                        waiting = container_status.state.waiting
                        if waiting is not None and waiting.reason != 'ContainerCreating':
                            all_pods_scheduled = False
                            break
```

I updated
```
                        if waiting and (waiting.reason == 'ErrImagePull' or
                                        waiting.reason == 'ImagePullBackOff'):
```
from `_wait_for_pods_to_run` with
```
                        if waiting is not None and waiting.reason != 'ContainerCreating':
```
so that the post-schedule errors can be hanlded from `_wait_for_pods_to_run`.